File: Algorithms Advanced/others/GraphAlgorithms/Python/Edmond-Karp.py
from collections import deque

def bfs(source, sink, residual, parent):
    visited = [False] * len(residual)
    queue = deque([source])
    visited[source] = True

    while queue:
        u = queue.popleft()
        for v in range(len(residual)):
            if not visited[v] and residual[u][v] > 0:
                parent[v] = u
                if v == sink:
                    return True
                visited[v] = True
                queue.append(v)
    return False

def edmond_karp(capacity, source, sink):
    nodes = len(capacity)
    residual = [row[:] for row in capacity]
    parent = [-1] * nodes
    max_flow = 0

    while bfs(source, sink, residual, parent):
        path_flow = float('inf')
        v = sink
        while v != source:
            u = parent[v]
            path_flow = min(path_flow, residual[u][v])
            v = parent[v]

        v = sink
        while v != source:
            u = parent[v]
            residual[u][v] -= path_flow
            residual[v][u] += path_flow
            v = parent[v]

        max_flow += path_flow

    return max_flow

# Example
capacity = [
    [0, 16, 13, 0, 0, 0],
    [0, 0, 10, 12, 0, 0],
    [0, 4, 0, 0, 14, 0],
    [0, 0, 9, 0, 0, 20],
    [0, 0, 0, 7, 0, 4],
    [0, 0, 0, 0, 0, 0]]
print("The maximum possible flow is", edmond_karp(capacity, 0, 5))

File: Algorithms Advanced/others/GraphAlgorithms/C++/Edmond-Karp.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

bool bfs(int source, int sink, vector<vector<int>> &residual, vector<int> &parent) 
{
    int nodes = residual.size();
    vector<bool> visited(nodes, false);
    queue<int> q;
    q.push(source);
    visited[source] = true;

    while (!q.empty()) 
    {
        int u = q.front();
        q.pop();

        for (int v = 0; v < nodes; v++) 
        {
            if (!visited[v] && residual[u][v] > 0) 
            {
                parent[v] = u;
                if (v == sink) return true;
                visited[v] = true;
                q.push(v);
            }
        }
    }
    return false;
}

int edmondKarp(vector<vector<int>> &capacity, int source, int sink) 
{
    int nodes = capacity.size();
    vector<vector<int>> residual = capacity;
    vector<int> parent(nodes, -1);
    int maxFlow = 0;

    while (bfs(source, sink, residual, parent)) 
    {
        int pathFlow = INT_MAX;
        for (int v = sink; v != source; v = parent[v]) 
        {
            int u = parent[v];
            pathFlow = min(pathFlow, residual[u][v]);
        }

        for (int v = sink; v != source; v = parent[v]) 
        {
            int u = parent[v];
            residual[u][v] -= pathFlow;
            residual[v][u] += pathFlow;
        }

        maxFlow += pathFlow;
    }

    return maxFlow;
}

int main() 
{
    vector<vector<int>> capacity = {
        {0, 16, 13, 0, 0, 0},
        {0, 0, 10, 12, 0, 0},
        {0, 4, 0, 0, 14, 0},
        {0, 0, 9, 0, 0, 20},
        {0, 0, 0, 7, 0, 4},
        {0, 0, 0, 0, 0, 0}};
    cout << "The maximum possible flow is " << edmondKarp(capacity, 0, 5) << endl;
    return 0;
}
